ch_tools/chadmin/internal/zookeeper.py

```diff
@@ -148,22 +143,63 @@ def _find_paths(zk, root_path, included_paths_regexp, excluded_paths_regexp=None
     return list(paths)
 
 
-def clean_zk_metadata_for_hosts(ctx, nodes):
+def _delete_nodes_transaction(zk, to_delete_in_trasaction):
     """
-    Perform cleanup in zookeeper after deleting hosts in the cluster or whole cluster deleting.
+    Perform deletion for the list of nodes in a single transaction.
+    If the transaction fails, go through the list and delete the nodes one by one.
     """
-
-    def _try_delete_zk_node(zk, node):
+    delete_transaction = zk.transaction()
+    for node in to_delete_in_trasaction:
+        delete_transaction.delete(node)
+    result = delete_transaction.commit()
+
+    if result.count(True) == len(result):
+        # Transaction completed successfully, exit.
+        return
+    for node in to_delete_in_trasaction:
         try:
-            _delete_zk_nodes(zk, [node])
+            zk.delete(node, recursive=True)
         except NoNodeError:
             #  Someone deleted node before us. Do nothing.
             print("Node {node} is already absent, skipped".format(node=node))
-        except NotEmptyError:
-            #  Someone created child node while deleting.
-            #  I'm not sure that we can get this exception with recursive=True.
-            #  Do nothing.
-            print("Node {node} is not empty, skipped".format(node=node))
+
+
+def _delete_recursive(zk, paths):
+    """
+    Kazoo already has the ability to recursively delete nodes, but the implementation is quite naive
+    and has poor performance with a large number of nodes being deleted.
+
+    In this implementation we unite the nodes to delete in transactions to do single operation for batch of nodes.
+    To delete in correct order first of all we perform topological sort using bfs approach.
+    """
+
+    print("Node to recursive delete", paths)
+    nodes_to_delete = []
+    queue: Queue = Queue()
```

Comment:
Actually, [`queue.Queue`](https://docs.python.org/3/library/queue.html) is intended for communication in threaded apps and has corresponding overhead. It is much more preferable to use [`collections.deque`](https://docs.python.org/3/library/collections.html?highlight=collections#collections.deque) as FIFO data structure.

It can be initialized as `d = deque(list_obj)`